unittests/ranges/test_ranges.cpp

```diff
@@ -0,0 +1,497 @@
+
+    manager = new QNetworkAccessManager(this);
+}
+
+void TestRangeRequests::cleanupTestCase()
+{
+    delete server;
+    delete manager;
+}
+
+void TestRangeRequests::testValidRanges_data()
+{
+    QTest::addColumn<QString>("rangeHeader");
+    QTest::addColumn<QList<QByteArray>>("expectedRanges");
+    QTest::addColumn<int>("expectedStatusCode");
+
+    QTest::newRow("full_range") << "bytes=0-2047"
+                                << QList<QByteArray> {expectedRange(0, 2048)} << 200; // Full file, 200 OK
+
+    QTest::newRow("suffix_equal_file") << "bytes=-2048"
+                                       << QList<QByteArray> {expectedRange(0, 2048)} << 200; // Full file suffix, 200 OK
+
+    QTest::newRow("suffix_larger_than_file") << "bytes=-9999"
+                                             << QList<QByteArray> {expectedRange(0, 2048)} << 200; // Suffix bigger than file, 200 OK
+
+    QTest::newRow("valid_single_range") << "bytes=100-199"
+                                        << QList<QByteArray> {expectedRange(100, 100)} << 206; // Partial content
+
+    QTest::newRow("multiple_ranges") << "bytes=100-199,300-399"
+                                     << QList<QByteArray> {expectedRange(100, 100), expectedRange(300, 100)} << 206; // Partial content
+
+    QTest::newRow("adjacent_coalesced") << "bytes=100-150,151-160"
+                                        << QList<QByteArray> {expectedRange(100, 61)} << 206; // Coalesced ranges
+
+    QTest::newRow("invalid_range_order") << "bytes=200-100"
+                                         << QList<QByteArray> {} << 416; // Unsatisfiable
+
+    QTest::newRow("unsatisfiable_range") << "bytes=3000-4000"
+                                         << QList<QByteArray> {} << 416; // Unsatisfiable
+}
+
+void TestRangeRequests::testValidRanges()
+{
+    QFETCH(QString, rangeHeader);
+    QFETCH(QList<QByteArray>, expectedRanges);
+    QFETCH(int, expectedStatusCode);
+
+    QNetworkRequest request(testUrl);
+    if (!rangeHeader.isEmpty())
+        request.setRawHeader("Range", rangeHeader.toUtf8());
+
+    QNetworkReply *reply = manager->get(request);
+    QSignalSpy spy(reply, &QNetworkReply::finished);
+    QVERIFY(spy.wait());
+
+    int status = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
+
+    if (expectedStatusCode != 200 || status == 200) // When transferring the whole file can either reply with 200 or 206
+        QCOMPARE(status, expectedStatusCode);
+    else
+        QCOMPARE(status, 206);
+
+    if (status == 206) {
+        QByteArray data = reply->readAll();
+        for (const QByteArray &range : expectedRanges) {
+            QVERIFY(data.contains(range));
+        }
+    } else if (status == 200) {
+        // For 200 OK full content, expect exact data match
+        QCOMPARE(reply->readAll(), testFileData);
+    } else {
+        // For 416 or others, body should be empty
+        QCOMPARE(reply->readAll().size(), 0);
+    }
+
+    reply->deleteLater();
```

Comment:
I think this could be `delete reply`, we're not in a slot connected to reply, and there's not really an event loop running to handle that deleteLater (they still all get deleted before exit, I think)

(repeats)